autouse yield_fixture doesn't seem to work for class members

* regular `autouse` fixture run correctly:
  ```python
import pytest
import unittest

class TestFoo(unittest.TestCase):
        @pytest.fixture(autouse=True)
        def set_flag(self, request):
            self.flag = True
            @request.addfinalizer
            def unset_flag():
                self.flag = False

        def test_foo(self):
            assert self.flag
```
  ```
> py.test -q test_foo.py
.
1 passed in 0.01 seconds
```

* a `yield_fixture` function *outside* the TestCase also works correctly:

  ```python
import pytest
import unittest

flag = False
@pytest.yield_fixture(autouse=True)
def set_flag():
        global flag
        flag = True
        yield
        flag = False

class TestFoo(unittest.TestCase):
        def test_foo(self):
            assert flag
```
  ```
> py.test -q test_foo.py
.
1 passed in 0.02 seconds
```

* However converting an autouse fixture method to a `yield_fixture` fails, the fixture doesn't seem to be run (the only modification from the first example is the conversion from fixture to yield_fixture)

  ```python
import pytest
import unittest

class TestFoo(unittest.TestCase):
        @pytest.yield_fixture(autouse=True)
        def set_flag(self):
            self.flag = True
            yield
            self.flag = False

        def test_foo(self):
            assert self.flag
```
  ```
> py.test -q test_foo.py
F
=============== FAILURES ===============
___________ TestFoo.test_foo ___________

self = <test_foo.TestFoo testMethod=test_foo>

    def test_foo(self):
>       assert flag
E       NameError: global name 'flag' is not defined

test_foo.py:13: NameError
1 failed in 0.03 seconds

```

--- a/CHANGELOG
+++ b/CHANGELOG
@@ -12,6 +12,10 @@
 - preserve warning functions after call to pytest.deprecated_call. Thanks
   Pieter Mulder for PR.
 
+- fix issue854: autouse yield_fixtures defined as class members of
+  unittest.TestCase subclasses now work as expected.
+  Thannks xmo-odoo for the report and Bruno Oliveira for the PR.
+
 - fix issue833: --fixtures now shows all fixtures of collected test files, instead of just the
   fixtures declared on the first one.
   Thanks Florian Bruhin for reporting and Bruno Oliveira for the PR.

--- a/_pytest/python.py
+++ b/_pytest/python.py
@@ -1900,10 +1900,13 @@ class FixtureDef:
             self.finish()
             assert not hasattr(self, "cached_result")
 
+        fixturefunc = self.func
+
         if self.unittest:
-            result = self.func(request.instance, **kwargs)
+            if request.instance is not None:
+                # bind the unbound method to the TestCase instance
+                fixturefunc = self.func.__get__(request.instance)
         else:
-            fixturefunc = self.func
             # the fixture function needs to be bound to the actual
             # request.instance so that code working with "self" behaves
             # as expected.
@@ -1911,12 +1914,13 @@ class FixtureDef:
                 fixturefunc = getimfunc(self.func)
                 if fixturefunc != self.func:
                     fixturefunc = fixturefunc.__get__(request.instance)
-            try:
-                result = call_fixture_func(fixturefunc, request, kwargs,
-                                           self.yieldctx)
-            except Exception:
-                self.cached_result = (None, my_cache_key, sys.exc_info())
-                raise
+
+        try:
+            result = call_fixture_func(fixturefunc, request, kwargs,
+                                       self.yieldctx)
+        except Exception:
+            self.cached_result = (None, my_cache_key, sys.exc_info())
+            raise
         self.cached_result = (result, my_cache_key, None)
         return result
 

--- a/testing/test_unittest.py
+++ b/testing/test_unittest.py
@@ -607,18 +607,23 @@ def test_unittest_unexpected_failure(testdir):
     ])
 
 
-
-def test_unittest_setup_interaction(testdir):
+@pytest.mark.parametrize('fix_type, stmt', [
+    ('fixture', 'return'),
+    ('yield_fixture', 'yield'),
+])
+def test_unittest_setup_interaction(testdir, fix_type, stmt):
     testdir.makepyfile("""
         import unittest
         import pytest
         class MyTestCase(unittest.TestCase):
-            @pytest.fixture(scope="class", autouse=True)
+            @pytest.{fix_type}(scope="class", autouse=True)
             def perclass(self, request):
                 request.cls.hello = "world"
-            @pytest.fixture(scope="function", autouse=True)
+                {stmt}
+            @pytest.{fix_type}(scope="function", autouse=True)
             def perfunction(self, request):
                 request.instance.funcname = request.function.__name__
+                {stmt}
 
             def test_method1(self):
                 assert self.funcname == "test_method1"
@@ -629,7 +634,7 @@ def test_unittest_setup_interaction(testdir):
 
             def test_classattr(self):
                 assert self.__class__.hello == "world"
-    """)
+    """.format(fix_type=fix_type, stmt=stmt))
     result = testdir.runpytest()
     result.stdout.fnmatch_lines("*3 passed*")
 

